### PR TITLE
Preview functionality for Apple News themes

### PR DESCRIPTION
**Pull request by bcampeau**

### CODE REVIEW

File: admin/class-admin-apple-preview.php [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+				);
+
+				// Get the order of the top components
+				$component_order = $settings->get( 'meta_component_order' );
+				foreach ( $component_order as $component ) {
+					echo wp_kses( $$component, Admin_Apple_Settings_Section::$allowed_html );
+				}
+			?>
+			<div class="apple-news-component">
+			<p><span class="apple-news-dropcap">L</span>orem ipsum dolor sit amet, consectetur adipiscing elit. Mauris sagittis, <a href="#">augue vitae iaculis euismod</a>, libero nulla pellentesque quam, non venenatis massa odio id dolor.</p>
```
Comment (romancandlethoughts) [inline admin/class-admin-apple-preview.php]: I assume this is place holder, but the markup indentation is off

File: admin/class-admin-apple-themes.php [hunk 2]
```diff
@@ -107,18 +125,55 @@ public function page_themes_render() {
...
+		// Check for a valid theme
+		if ( ! isset( $_GET['theme'] ) ) {
+			$error = __( 'No theme was specified to preview', 'apple-news' );
+		} else {
+			$theme_name = sanitize_text_field( $_GET['theme'] );
+
+			// Load the theme
+			$theme = get_option( $this->theme_key_from_name( $theme_name ) );
+			if ( empty( $theme ) || ! is_array( $theme ) ) {
+				$error = sprintf(
```
Comment (romancandlethoughts) [inline admin/class-admin-apple-themes.php]: did you need to do anything with `$error` in this context? Also see L140 above.

File: admin/class-admin-apple-themes.php [hunk 3]
```diff
@@ -107,18 +125,55 @@ public function page_themes_render() {
...
 	 * Register assets for the options page.
 	 *
 	 * @param string $hook
 	 * @access public
 	 */
 	public function register_assets( $hook ) {
-		if ( 'apple-news_page_apple-news-themes' != $hook ) {
+		if ( ! in_array( $hook, array(
+			'apple-news_page_apple-news-themes',
+			'apple-news_page_apple-news-theme-preview'
```
Comment (romancandlethoughts) [inline admin/class-admin-apple-themes.php]: comma

File: admin/class-admin-apple-themes.php [hunk 4]
```diff
@@ -107,18 +125,55 @@ public function page_themes_render() {
...
+
+	/**
 	 * Register assets for the options page.
 	 *
 	 * @param string $hook
 	 * @access public
 	 */
 	public function register_assets( $hook ) {
-		if ( 'apple-news_page_apple-news-themes' != $hook ) {
+		if ( ! in_array( $hook, array(
```
Comment (romancandlethoughts) [inline admin/class-admin-apple-themes.php]: strict if needed

File: admin/class-admin-apple-themes.php [hunk 5]
```diff
@@ -135,30 +190,28 @@ public function register_assets( $hook ) {
 	 * List all available themes
 	 *
 	 * @access public
-	 * @static
 	 */
```
Comment (romancandlethoughts) [inline admin/class-admin-apple-themes.php]: no `@return`

File: admin/class-admin-apple-themes.php [hunk 6]
```diff
@@ -135,30 +190,28 @@ public function register_assets( $hook ) {
...
 		return get_option( self::theme_active_key );
 	}
 
 	/**
 	 * Get a specific theme
 	 *
+	 * @param string $name
 	 * @access public
-	 * @static
 	 */
```
Comment (romancandlethoughts) [inline admin/class-admin-apple-themes.php]: no `@return`

File: admin/class-admin-apple-themes.php [hunk 7]
```diff
@@ -616,9 +669,38 @@ private function validate_data( $data ) {
...
+	 *
+	 * @param string $name
+	 * @return string
+	 * @access public
+	 */
+	public function theme_preview_url( $name ) {
+		return add_query_arg(
+			array(
+				'page' => $this->theme_preview_page_name,
+				'theme' => $name
```
Comment (romancandlethoughts) [inline admin/class-admin-apple-themes.php]: comma

File: admin/partials/page_theme_preview.php [hunk 8]
```diff
@@ -0,0 +1,13 @@
+<div class="wrap apple-news-theme-preview">
+	<?php if ( ! empty( $error ) ) : ?>
+	<p class="error-message"><?php echo esc_html( $error ) ?></p>
```
Comment (romancandlethoughts) [inline admin/partials/page_theme_preview.php]: indent

File: admin/partials/page_theme_preview.php [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+	<?php else : ?>
+	<h1 id="apple_news_themes_title"><?php esc_html_e( 'Previewing', 'apple-news' ) ?> <?php echo esc_html( $theme_name ) ?></h1>
```
Comment (romancandlethoughts) [inline admin/partials/page_theme_preview.php]: indentation is off here too

File: admin/partials/page_themes.php [hunk 10]
```diff
@@ -68,32 +68,34 @@
...
 			<?php
-				$themes = \Admin_Apple_Themes::list_themes();
-				$active_theme = \Admin_Apple_Themes::get_active_theme();
-				if ( empty( $themes ) ) :
+				$themes = new \Admin_Apple_Themes();
+				$all_themes = $themes->list_themes();
+				$active_theme = $themes->get_active_theme();
+				if ( empty( $all_themes ) ) :
 				?>
 				<tr>
```
Comment (romancandlethoughts) [inline admin/partials/page_themes.php]: indentation is off

File: admin/partials/page_themes.php [hunk 11]
```diff
@@ -68,32 +68,34 @@
...
+							</th>
+							<td class="name column-apple-news-theme-name column-primary" data-colname="Name">
+								<?php echo esc_html( $theme ) ?>
+								<button type="button" class="toggle-row"><span class="screen-reader-text"><?php esc_html_e( 'Show more details', 'apple-news' ) ?></span></button>
+							</td>
+							<td class="column-apple-news-theme-actions" data-colname="Actions">
+								<a href="#" class="apple-news-row-action apple-news-export-theme" data-theme="<?php echo esc_attr( $theme ) ?>"><?php esc_html_e( 'Export', 'apple-news' ) ?></a>
+								<a href="<?php echo esc_url( $themes->theme_preview_url( $theme ) ) ?>" class="apple-news-row-action apple-news-preview-theme" data-theme="<?php echo esc_attr( $theme ) ?>"><?php esc_html_e( 'Preview', 'apple-news' ) ?></a>
+								<?php if ( $theme !== $active_theme ) : ?>
+								<a href="#" class="apple-news-row-action apple-news-delete-theme" data-theme="<?php echo esc_attr( $theme ) ?>"><?php esc_html_e( 'Delete', 'apple-news' ) ?></a>
```
Comment (romancandlethoughts) [inline admin/partials/page_themes.php]: indent

File: assets/js/preview.js [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+		// Do an initial update
+		appleNewsUpdatePreview();
+
+		// Check that we are fully compatible
+		if ( ! appleNewsSupportsMacFeatures() ) {
+			$( '.apple-news-settings-preview' ).prepend(
+				$( '<div>' )
+					.addClass( 'font-notice' )
+					.text( appleNewsSettings.fontNotice )
+			)
```
Comment (romancandlethoughts) [inline assets/js/preview.js]: missing semi colon

File: assets/js/preview.js [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+			value = value + units;
+		}
+
+		$( displayElement ).css( property, value );
+	}
+
+}( jQuery ) );
+
+function appleNewsSupportsMacFeatures() {
+	if ( 'MacIntel' === navigator.platform ) {
```
Comment (romancandlethoughts) [inline assets/js/preview.js]: is `'MacIntel'` exhaustive for all mac platforms that could support mac features?

File: admin/class-admin-apple-preview.php [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+	}
+
+	/**
+	 * Register assets for the options page.
+	 *
+	 * @param string $hook
+	 * @access public
+	 */
+	public function register_assets( $hook ) {
+		if ( ! in_array( $hook, array(
```
Comment (bcampeau) [inline admin/class-admin-apple-preview.php]: Sure

File: admin/class-admin-apple-preview.php [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+			'apple-news_page_apple-news-options',
+			'admin_page_apple-news-theme-preview'
```
Comment (bcampeau) [inline admin/class-admin-apple-preview.php]: OK

File: admin/class-admin-apple-preview.php [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+		) ) ) {
+			return;
+		}
+
+		wp_enqueue_style( 'apple-news-preview-css', plugin_dir_url( __FILE__ ) .
```
Comment (bcampeau) [inline admin/class-admin-apple-preview.php]: This is a convention across the plugin, so maybe we'll clean it up in a future version.

File: admin/class-admin-apple-preview.php [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+		$admin_settings = new Admin_Apple_Settings();
+		$settings = $admin_settings->fetch_settings();
+
+		// If a theme name is provided, replace formatting settings with those from the theme.
+		if ( ! empty( $theme ) ) {
+			$themes = new Admin_Apple_Themes();
+			$theme_settings = $themes->get_theme( $theme );
+			if ( empty( $theme_settings ) || ! is_array( $theme_settings ) ) {
+				?>
+				<p class="error-message"><?php echo sprintf(
```
Comment (bcampeau) [inline admin/class-admin-apple-preview.php]: Too early. It's escaped when output later on.

File: admin/class-admin-apple-preview.php [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+			// If true, add hidden form fields for the settings.
+			// This is used for theme preview.
+			if ( $add_settings ) :
+			?>
+				<form id="apple-news-settings-form">
+			<?php
+			endif;
+
+			// Replace all the formatting settings
+			foreach ( $theme_settings as $key => $value ) {
```
Comment (bcampeau) [inline admin/class-admin-apple-preview.php]: `$theme_settings` is validated on line 61. The function would exit if so.

File: admin/class-admin-apple-preview.php [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+				$export = new Apple_Actions\Index\Export( $settings );
+				$byline = sprintf(
+					'<div class="apple-news-byline apple-news-component apple-news-meta-component">%s</div>',
+					$export->format_byline( null, $author, $date )
+				);
+
+				// Get the order of the top components
+				$component_order = $settings->get( 'meta_component_order' );
+				foreach ( $component_order as $component ) {
+					echo wp_kses( $$component, Admin_Apple_Settings_Section::$allowed_html );
```
Comment (bcampeau) [inline admin/class-admin-apple-preview.php]: Yes, intentional

File: admin/class-admin-apple-preview.php (hunk 1 above)
Comment (bcampeau) [inline admin/class-admin-apple-preview.php]: This provides content for the preview mode for a theme or settings.

File: admin/class-admin-apple-themes.php (hunk 2 above)
Comment (bcampeau) [inline admin/class-admin-apple-themes.php]: The `$error` variable is used on the actual template page referenced on line 158.

File: admin/class-admin-apple-themes.php (hunk 4 above)
Comment (bcampeau) [inline admin/class-admin-apple-themes.php]: ok

File: admin/class-admin-apple-themes.php (hunk 3 above)
Comment (bcampeau) [inline admin/class-admin-apple-themes.php]: ok

File: admin/class-admin-apple-themes.php (hunk 5 above)
Comment (bcampeau) [inline admin/class-admin-apple-themes.php]: will fix

File: admin/class-admin-apple-themes.php (hunk 6 above)
Comment (bcampeau) [inline admin/class-admin-apple-themes.php]: will fix

File: admin/class-admin-apple-themes.php (hunk 7 above)
Comment (bcampeau) [inline admin/class-admin-apple-themes.php]: will fix

File: admin/partials/page_theme_preview.php (hunk 8 above)
Comment (bcampeau) [inline admin/partials/page_theme_preview.php]: ok

File: admin/partials/page_theme_preview.php (hunk 9 above)
Comment (bcampeau) [inline admin/partials/page_theme_preview.php]: ok

File: admin/partials/page_themes.php (hunk 10 above)
Comment (bcampeau) [inline admin/partials/page_themes.php]: I guess

File: admin/partials/page_themes.php (hunk 11 above)
Comment (bcampeau) [inline admin/partials/page_themes.php]: ok

File: assets/js/preview.js (hunk 12 above)
Comment (bcampeau) [inline assets/js/preview.js]: ok

File: assets/js/preview.js (hunk 13 above)
Comment (bcampeau) [inline assets/js/preview.js]: Yes, this question has been asked before. Any other Mac platforms would not support the fonts required.

**Pull request merged**